### PR TITLE
Add FIPS selftest and get_result_selftest

### PR DESCRIPTION
**Pull request by Wenxing-hou**
reference: #1260

**Commit 1:  Add HMAC self_test;
Commit 2:  Add AES_GCM self_test;
Commit 3:  Add RSA_SSA self_test;**

When the LIBSPDM_FIPS_MODE is open, the test result:
image

When the LIBSPDM_FIPS_MODE is closed, the test result:
image

Signed-off-by: Wenxing Hou <wenxing.hou@intel.com>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -848,6 +848,7 @@ else()
         ADD_SUBDIRECTORY(os_stub/spdm_device_secret_lib_null)
         ADD_SUBDIRECTORY(unit_test/spdm_transport_test_lib)
         ADD_SUBDIRECTORY(unit_test/cmockalib)
+        ADD_SUBDIRECTORY(library/spdm_crypt_lib/fips)
```
Comment (steven-bellock) [inline CMakeLists.txt:851]: Does this need to be a separate library? Why not make it part of `spdm_crypt_lib`?

File: include/internal/libspdm_fips_lib.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/**
+ *  Copyright Notice:
+ *  Copyright 2021-2022 DMTF. All rights reserved.
```
Comment (steven-bellock) [inline include/internal/libspdm_fips_lib.h:3]: "2021-2022" -> "2023"

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ *  Copyright Notice:
+ *  Copyright 2021-2022 DMTF. All rights reserved.
+ *  License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/libspdm/blob/main/LICENSE.md
+ **/
+
+#include "internal/libspdm_crypt_lib.h"
+
+#if LIBSPDM_FIPS_MODE
+
+static const uint8_t key[32] = {0};
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:11]: These can all be local to the self-test function. No need to statically allocate them.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+/**
+ * HMAC KAT test: HMAC-SHA512 KAT and HMAC-SHA256 KAT.
+ * HMAC-SHA512 KAT covers SHA384, SHA512, HMAC-SHA384, and HMAC-SHA512.
+ * HMAC-SHA256 KAT covers SHA1, SHA224, SHA256, HMAC-SHA1, HMAC-SHA224, and HMAC-SHA256.
+ **/
+bool libspdm_fips_selftest_hmac(void)
+{
+    uint8_t hmac_256_result[32];
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:40]: The result needs to be zeroized before calling `libspdm_hmac_sha256_all`.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    uint8_t hmac_512_result[64];
+    bool result;
+    result = true;
+
+#if LIBSPDM_SHA256_SUPPORT
+    result = libspdm_hmac_sha256_all(msg, sizeof(msg), key, sizeof(key), hmac_256_result);
+    if (!result) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "gen hmac_sha256 failed \n"));
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:48]: Which algorithm failed needs to be conveyed to the caller, and it should probably not return early. For example if SHA-256 fails but SHA-512 passes then Integrator can use SHA-512 but not SHA-256.

**Comment by jyao1** [comment]
for selftest location, please refer to https://github.com/DMTF/libspdm/discussions/1406#discussioncomment-4712827

File: CMakeLists.txt (hunk 1 above)
Comment (Wenxing-hou) [reply CMakeLists.txt]: Thanks. I have included the fips as the part of `spdm_crypt_lib`.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c (hunk 3 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c]: Thanks. I have move them to local.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c (hunk 4 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c]: Thanks. I have added the zero_mem.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c (hunk 5 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c]: I have added the `status` as the final result.

**Comment by jyao1** [comment]
Please add all sym, deterministic asym algo at first.
Then probabilistic asym algo with hardcoded random.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    if (libspdm_const_compare_mem(hmac_256_result, hmac_sha256_answer,
+                                  sizeof(hmac_sha256_answer))) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "hmac_sha256 KAT failed \n"));
+        status = false;
+    }
+
+#endif/*LIBSPDM_SHA256_SUPPORT*/
+
+#if LIBSPDM_SHA512_SUPPORT
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:52]: need add LIBSPDM_SHA384_SUPPORT

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        0xfb, 0x9e, 0x8a, 0xa1, 0x1a, 0x8d, 0xe9, 0xa4,
+        0xea, 0x15, 0xa0, 0x45, 0xf2, 0xbe, 0x24, 0x5f,
+        0xbb, 0x98, 0x91, 0x6a, 0x9a, 0xe8, 0x1b, 0x35,
+        0x3e, 0x33, 0xb9, 0xc4, 0x2a, 0x55, 0x38, 0x0c,
+        0x51, 0x58, 0x24, 0x1d, 0xae, 0xb3, 0xc6, 0xdd
+    };
+    libspdm_zero_mem(hmac_512_result, sizeof(hmac_512_result));
+    result = libspdm_hmac_sha512_all(msg, sizeof(msg), key, sizeof(key), hmac_512_result);
+    if (!result) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "gen hmac_sha384 failed \n"));
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:67]: should be hmac_sha512?

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        status = false;
+    }
+
+    if (libspdm_const_compare_mem(hmac_512_result, hmac_sha512_answer,
+                                  sizeof(hmac_sha512_answer))) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "hmac_sha384 KAT failed \n"));
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:73]: should be hmac_sha512?

File: library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+ *  Copyright 2023 DMTF. All rights reserved.
+ *  License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/libspdm/blob/main/LICENSE.md
+ **/
+
+#include "internal/libspdm_crypt_lib.h"
+
+#if LIBSPDM_FIPS_MODE
+
+/**
+ * AES_GCM self_test: AES_GCM KAT test and AES_GCM PCT test.
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c:12]: This is confusing - please remove the comment for "PCT test".

File: library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "aes_gcm KAT failed \n"));
+        return false;
+    }
+
+    if (libspdm_const_compare_mem(output_tag, expected_tag, sizeof(expected_tag))) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "aes_gcm KAT failed \n"));
+        return false;
+    }
+
+    /*PCT test*/
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c:120]: This is confusing - please remove the comment for "PCT test".

File: library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+ *  Copyright 2023 DMTF. All rights reserved.
+ *  License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/libspdm/blob/main/LICENSE.md
+ **/
+
+#include "internal/libspdm_crypt_lib.h"
+
+#if LIBSPDM_FIPS_MODE
+
+/**
+ * RSA_SSA(RSASSA-PKCS1 v1.5) self_test: RSA_SSA KAT test and RSA_SSA PCT test.
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c:12]: This is confusing - please remove the comment for "PCT test".

File: library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    if (libspdm_const_compare_mem(signature, expected_sig,
+                                  sizeof(expected_sig))) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "RSA_PSS KAT failed \n"));
+        libspdm_rsa_free(rsa);
+        return false;
+    }
+
+    /*PCT test*/
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c:193]: This is confusing - please remove the comment for "PCT test".

**Comment by jyao1** [comment]
This is enough as the first patch.

I recommend, we review the design at first, then merge, then continue with rest key exchange or probabilistic algo.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c (hunk 7 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c]: Thanks. I have changed the code.

File: library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c (hunk 9 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c]: Thanks. I have deleted the "PCT test" comment.

File: library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                                          expected_ciphertext, sizeof(expected_ciphertext),
+                                          expected_tag, sizeof(expected_tag),
+                                          output_plaintext, &output_plaintext_size);
+    if (!result) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "aes_gcm decrypt failed \n"));
+        return false;
+    }
+
+    if (output_plaintext_size != sizeof(input)) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "aes_gcm PCT failed \n"));
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c:131]: remove PCT

File: library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        return false;
+    }
+
+    if (libspdm_const_compare_mem(output_plaintext, input, sizeof(input))) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "aes_gcm PCT failed \n"));
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c:136]: remove PCT

File: library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "RSA_PSS KAT failed \n"));
+        libspdm_rsa_free(rsa);
+        return false;
+    }
+
+    result = libspdm_rsa_pkcs1_verify_with_nid(rsa, LIBSPDM_CRYPTO_NID_SHA256,
+                                               input_data, sizeof(input_data),
+                                               signature, sig_size);
+    if (!result) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "RSA_PSS PCT failed \n"));
```
Comment (jyao1) [inline library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c:197]: remove PCT

Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c]: Thanks. I have changed the `PCT` to `selftest`.

File: include/internal/libspdm_fips_lib.h [hunk 16]
```diff
@@ -0,0 +1,26 @@
+/**
+ *  Copyright Notice:
+ *  Copyright 2023 DMTF. All rights reserved.
+ *  License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/libspdm/blob/main/LICENSE.md
+ **/
+
+#if LIBSPDM_FIPS_MODE
+
+/**
+ * HMAC KAT test: HMAC-SHA512 KAT and HMAC-SHA256 KAT.
```
Comment (jyao1) [inline include/internal/libspdm_fips_lib.h:10]: need add HMAC-SHA384

Comment (Wenxing-hou) [reply include/internal/libspdm_fips_lib.h]: Thanks. I have added the HMAC-SHA384 KAT.

File: include/internal/libspdm_fips_lib.h [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+ *  License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/libspdm/blob/main/LICENSE.md
+ **/
+
+#if LIBSPDM_FIPS_MODE
+
+/**
+ * HMAC KAT test: HMAC-SHA512 KAT and HMAC-SHA384 KAT and HMAC-SHA256 KAT.
+ * HMAC-SHA512 KAT covers SHA384, SHA512, HMAC-SHA384, and HMAC-SHA512.
+ * HMAC-SHA384 KAT covers SHA384, SHA512, HMAC-SHA384, and HMAC-SHA512.
+ * HMAC-SHA256 KAT covers SHA1, SHA224, SHA256, HMAC-SHA1, HMAC-SHA224, and HMAC-SHA256.
```
Comment (steven-bellock) [inline include/internal/libspdm_fips_lib.h:13]: Only list the algorithms that will be covered. For example SPDM doesn't support SHA1.

File: library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+    uint8_t output_plaintext[1024];
+    uint8_t output_tag[1024];
+    size_t output_ciphertext_size;
+    size_t output_plaintext_size;
+
+    output_ciphertext_size = sizeof(output_ciphertext);
+    output_plaintext_size = sizeof(output_plaintext);
+
+    /*Key to use crypto operation*/
+    const uint8_t key[16] = {
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c:30]: You can get rid of the explicit array sizes.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+    if (libspdm_const_compare_mem(hmac_256_result, hmac_sha256_answer,
+                                  sizeof(hmac_sha256_answer))) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "hmac_sha256 KAT failed \n"));
+        status = false;
+    }
+
+#endif/*LIBSPDM_SHA256_SUPPORT*/
+
+#if LIBSPDM_SHA384_SUPPORT
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:53]: Make this a separate function.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+    if (libspdm_const_compare_mem(hmac_384_result, hmac_sha384_answer,
+                                  sizeof(hmac_sha384_answer))) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "hmac_sha384 KAT failed \n"));
+        status = false;
+    }
+
+#endif/*LIBSPDM_SHA384_SUPPORT*/
+
+#if LIBSPDM_SHA512_SUPPORT
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:78]: Make this a separate function.

File: library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+
+#if LIBSPDM_FIPS_MODE
+
+/**
+ * RSA_SSA(RSASSA-PKCS1 v1.5) self_test
+ **/
+bool libspdm_fips_selftest_rsa_ssa(void)
+{
+    bool result;
+    result = true;
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c:17]: For C89 compatibility we don't allow procedural statements until all variables have been declared for a scope. So I would make this `bool result = true;`.

**Review by steven-bellock — CHANGES_REQUESTED** [review]
@jyao1 Is this level of granularity ok? For example, if SHA-256 passes but SHA-384 fails, would you like to know that information? Or is any failure deemed fatal?

File: include/internal/libspdm_fips_lib.h (hunk 17 above)
Comment (Wenxing-hou) [reply include/internal/libspdm_fips_lib.h]: Thanks. I have made the comments more clear.

File: library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c (hunk 18 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_aes_gcm.c]: Yes. I deleted all explicit array sizes.

File: library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c (hunk 19 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_hmac.c:54]: Thanks. I have made three separate function.

File: library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c (hunk 21 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/fips/libspdm_selftest_rsa_ssa.c]: Thanks. I will remember it.

**Review by steven-bellock — APPROVED** [review]

**Pull request merged**